flink-runtime/src/test/java/org/apache/flink/runtime/asyncprocessing/AsyncExecutionControllerTest.java

```diff
@@ -346,6 +337,63 @@ public void testSyncPoint() {
         recordContext2.release();
     }
 
+    @Test
+    void testBufferTimeout() throws InterruptedException {
+        batchSize = 5;
+        timeout = 1000;
+        setup();
+        Runnable userCode = () -> valueState.asyncValue();
+
+        // ------------ basic timeout -------------------
+        for (int i = 0; i < batchSize - 1; i++) {
+            String record = String.format("key%d-r%d", i, i);
+            String key = String.format("key%d", batchSize + i);
+            RecordContext<String> recordContext = aec.buildContext(record, key);
+            aec.setCurrentContext(recordContext);
+            userCode.run();
+        }
+        assertThat(aec.timeoutFlag.get()).isFalse();
+        assertThat(aec.currentScheduledFuture.isDone()).isFalse();
+        assertThat(aec.inFlightRecordNum.get()).isEqualTo(batchSize - 1);
+        assertThat(aec.stateRequestsBuffer.activeQueueSize()).isEqualTo(batchSize - 1);
+        assertThat(aec.stateRequestsBuffer.blockingQueueSize()).isEqualTo(0);
+
+        Thread.sleep(timeout + 100);
```

Comment:
It might be better to avoid using `Thread.sleep` in test cases, as it may improve the duration of CI and behave as flaky tests. How about introduce a `TestScheduledThreadPoolExecutor` so we can control when each step is triggered? An example of this is `org.apache.flink.util.concurrent.ManuallyTriggeredScheduledExecutor`.

Reply:
Thanks for the suggestion, removed  `Thread.sleep` from `testBufferTimeout()` and `testBufferTimeoutSkip()`, used `ManuallyTriggeredScheduledExecutorService` as `ScheduledExecutor`.